File: android/app/src/main/kotlin/com/goalmaker/app/domain/update/ReleaseManifestParser.kt
package com.goalmaker.app.domain.update

import com.goalmaker.app.domain.version.SemanticVersion
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.longOrNull

/**
 * Parses and validates release manifest bytes that already passed the signature check. Unknown
 * properties are ignored; every rule here is covered by contracts/vectors/release-manifest.json.
 */
object ReleaseManifestParser {
    private const val SUPPORTED_SCHEMA = 1
    private const val MAX_ARTIFACT_SIZE = 524_288_000L
    private val pathPattern = Regex("^[A-Za-z0-9._-]+(/[A-Za-z0-9._-]+)*$")
    private val sha256Pattern = Regex("^[0-9a-f]{64}$")

    fun parse(bytes: ByteArray): ManifestCheck {
        val root = try {
            Json.parseToJsonElement(bytes.decodeToString()) as? JsonObject
        } catch (_: SerializationException) {
            null
        } ?: return ManifestCheck.BadManifest("not a JSON object")

        val schema = root.primitive("schema")?.intOrNull
        if (schema != SUPPORTED_SCHEMA) return ManifestCheck.BadManifest("unsupported schema $schema")

        val versionText = root.text("version") ?: return ManifestCheck.BadManifest("version missing")
        val version = SemanticVersion.parse(versionText)
            ?: return ManifestCheck.BadManifest("version is not semantic: $versionText")

        val publishedAt = root.text("publishedAt") ?: return ManifestCheck.BadManifest("publishedAt missing")

        val artifactsJson = root["artifacts"] as? JsonArray
        if (artifactsJson.isNullOrEmpty()) return ManifestCheck.BadManifest("no artifacts")
        val artifacts = artifactsJson.map { element ->
            parseArtifact(element) ?: return ManifestCheck.BadManifest("invalid artifact $element")
        }

        val androidVersionCode = root.primitive("androidVersionCode")?.intOrNull?.takeIf { it >= 1 }
        if (artifacts.any { it.platform == ReleasePlatform.ANDROID } && androidVersionCode == null) {
            return ManifestCheck.BadManifest("android artifact without androidVersionCode")
        }

        return ManifestCheck.Valid(
            ReleaseManifest(
                version = version,
                androidVersionCode = androidVersionCode,
                publishedAt = publishedAt,
                notes = root.text("notes"),
                artifacts = artifacts,
            ),
        )
    }

    private fun parseArtifact(element: JsonElement): ReleaseArtifact? {
        val item = element as? JsonObject ?: return null
        val platform = item.text("platform")?.let(ReleasePlatform::fromWireName) ?: return null
        val path = item.text("path")?.takeIf { value ->
            pathPattern.matches(value) && value.split('/').none { it == ".." || it == "." }
        } ?: return null
        val size = item.primitive("size")?.takeUnless { it.isString }?.longOrNull
            ?.takeIf { it in 1..MAX_ARTIFACT_SIZE } ?: return null
        val sha256 = item.text("sha256")?.takeIf(sha256Pattern::matches) ?: return null
        return ReleaseArtifact(platform, path, size, sha256)
    }

    private fun JsonObject.primitive(name: String): JsonPrimitive? = this[name] as? JsonPrimitive

    private fun JsonObject.text(name: String): String? = primitive(name)?.takeIf { it.isString }?.contentOrNull
}

File: windows/src/GoalMaker.Core/Updates/ReleaseManifestParser.cs
using System.Text.Json;
using System.Text.RegularExpressions;
using GoalMaker.Core.Versioning;

namespace GoalMaker.Core.Updates;

/// <summary>
/// Parses and validates release manifest bytes that already passed the signature check. Unknown
/// properties are ignored; every rule is covered by contracts/vectors/release-manifest.json.
/// </summary>
public static partial class ReleaseManifestParser
{
    private const int SupportedSchema = 1;
    private const long MaxArtifactSize = 524_288_000;

    public static ManifestCheck Parse(ReadOnlyMemory<byte> bytes)
    {
        JsonDocument document;
        try
        {
            document = JsonDocument.Parse(bytes);
        }
        catch (JsonException)
        {
            return new ManifestCheck.BadManifest("not JSON");
        }

        using (document)
        {
            var root = document.RootElement;
            if (root.ValueKind != JsonValueKind.Object)
            {
                return new ManifestCheck.BadManifest("not a JSON object");
            }

            if (Integer(root, "schema") != SupportedSchema)
            {
                return new ManifestCheck.BadManifest("unsupported schema");
            }

            var versionText = Text(root, "version");
            var version = versionText is null ? null : SemanticVersion.Parse(versionText);
            if (version is null)
            {
                return new ManifestCheck.BadManifest($"version is not semantic: {versionText}");
            }

            var publishedAt = Text(root, "publishedAt");
            if (publishedAt is null)
            {
                return new ManifestCheck.BadManifest("publishedAt missing");
            }

            if (!root.TryGetProperty("artifacts", out var artifactsJson)
                || artifactsJson.ValueKind != JsonValueKind.Array
                || artifactsJson.GetArrayLength() == 0)
            {
                return new ManifestCheck.BadManifest("no artifacts");
            }

            var artifacts = new List<ReleaseArtifact>();
            foreach (var element in artifactsJson.EnumerateArray())
            {
                var artifact = ParseArtifact(element);
                if (artifact is null)
                {
                    return new ManifestCheck.BadManifest($"invalid artifact {element.GetRawText()}");
                }

                artifacts.Add(artifact);
            }

            var androidVersionCode = Integer(root, "androidVersionCode") is { } code && code >= 1 ? (int?)code : null;
            if (androidVersionCode is null && artifacts.Any(artifact => artifact.Platform == ReleasePlatform.Android))
            {
                return new ManifestCheck.BadManifest("android artifact without androidVersionCode");
            }

            return new ManifestCheck.Valid(
                new ReleaseManifest(version, androidVersionCode, publishedAt, Text(root, "notes"), artifacts));
        }
    }

    private static ReleaseArtifact? ParseArtifact(JsonElement element)
    {
        if (element.ValueKind != JsonValueKind.Object)
        {
            return null;
        }

        ReleasePlatform? platform = Text(element, "platform") switch
        {
            "android" => ReleasePlatform.Android,
            "windows" => ReleasePlatform.Windows,
            _ => null,
        };
        var path = Text(element, "path");
        var sha256 = Text(element, "sha256");
        if (!element.TryGetProperty("size", out var sizeJson)
            || sizeJson.ValueKind != JsonValueKind.Number
            || !sizeJson.TryGetInt64(out var size))
        {
            return null;
        }

        var pathValid = path is not null
            && PathPattern().IsMatch(path)
            && !path.Split('/').Any(segment => segment is ".." or ".");
        if (platform is null || !pathValid || size is < 1 or > MaxArtifactSize
            || sha256 is null || !Sha256Pattern().IsMatch(sha256))
        {
            return null;
        }

        return new ReleaseArtifact(platform.Value, path!, size, sha256);
    }

    private static string? Text(JsonElement element, string name) =>
        element.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String ? value.GetString() : null;

    private static long? Integer(JsonElement element, string name) =>
        element.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.Number && value.TryGetInt64(out var number)
            ? number
            : null;

    // \z, not $: in .NET, $ also matches before a trailing newline.
    [GeneratedRegex(@"^[A-Za-z0-9._-]+(/[A-Za-z0-9._-]+)*\z", RegexOptions.CultureInvariant)]
    private static partial Regex PathPattern();

    [GeneratedRegex(@"^[0-9a-f]{64}\z", RegexOptions.CultureInvariant)]
    private static partial Regex Sha256Pattern();
}
